### PR TITLE
Match Markdown relative images and replace them with data URIs

### PR DESCRIPTION
**Pull request by stevehollaar**

### CODE REVIEW

**Comment by idris** [comment]
A few things:
1. Perhaps we should add abstractions for value => html rendering, so that it'll get configured with a map of extensions => processors.  Markdown would be one of those, and would run it through `marked`.
2. Perhaps the replacement of relative images (or file references) should also be configurable, so that we can use data-URIs or http URLs that will be hosted on a CDN.  This would be orthogonal to the markdown processor because it would work for other formats like textile, html, etc.  It would be a simple API of `(filePath) => Promise<fileUrl>`
3. I thought we decided against data-URIs for our internal use.

File: src/stores/GithubStore.js [hunk 1]
```diff
@@ -21,22 +73,13 @@ export default class GithubStore {
...
-    const pathParts = key.split('.');
-    if (this.config.rootPath) {
-      pathParts.unshift(this.config.rootPath);
-    }
-    let path = pathParts.join('/');
-    if (this.config.fileSuffix) {
-      path += '.md';
-    }
-
+  async getValueFromPath(objectPath, decode = true) {
```
Comment (idris) [inline src/stores/GithubStore.js]: Two things here:
1. Should this return a `Stream`? Or does Buffer also avoid forcing the entire file into memory?
2. Maybe this should be named something like `streamBinaryFromPath` to make clear that this is for getting raw binary data from the store.

File: src/stores/GithubStore.js [hunk 2]
```diff
@@ -46,15 +45,41 @@ export default class GithubStore {
...
 
-    if (response.content) {
-      return Buffer.from(response.content, 'base64').toString().trim();
+    let { content } = response;
+
+    if (content) {
+      content = (decode ? Buffer.from(content, 'base64').toString() : content).trim();
+
+      if (render) {
+        return this.renderer.render(content, {
```
Comment (idris) [inline src/stores/GithubStore.js]: I think we can keep this outside of the store, and get rid of the `selfRenders` option.  We always know the `contentPath` outside of the store, and the store can still expose `getContentFromPath`.

That way we keep rendering completely independent from store.

**Review by idris — CHANGES_REQUESTED** [review]
@stevehollaar see https://github.com/Truebill/smcms/pull/3#r78640019

**Comment by stevehollaar** [comment]
### Planned changes

**Content paths:**
- `/` separated
- include file extensions

**SMCS:**
- instantiated with:
  - `getRenderer` method (optional)
    - defaults to map of file extensions to renderers (e.g. `.md` -> `new MarkdownRenderer()`), with fallback to no-op renderer
  - `store` instance (required)
  - `resolveRelativePath` method (optional)
    - default to `path.join(path.dirname(contentPath), relativePath)`

**Stores**
- must expose `getValue` method
- ~~`getValue` must return a stream~~

**Renderers**
- must expose `render` method
- ~~`render` method operates on streams~~

**Comment by idris** [comment]
👍
My comments:
- maybe just rename `resolveRelativePaths` to `resolveRelativePath` (singular)
- If we're not doing base64 stuff, the stream stuff becomes less necessary.  We can push that off to a later version.

**Review by idris — APPROVED** [review]

**Comment by idris** [comment]
@stevehollaar merged.  Once you have it working in billninja, we can release a new version to npm

**Pull request merged**